drift_detector.go

```diff
@@ -2,57 +2,142 @@ package main
 	for range ticker.C {
 		detector.Validate()
 	}
 }
 
+// Validate queries the CF API and validates responses against the Watchtower config.
+// Results of (non-)compliance are exported as prometheus metrics via the /metrics endpoint.
 func (detector *Detector) Validate() {
-	detector.validateApps()
+	// Parallelize calls to validateX using goroutines and a sync.WaitGroup
+	var waitgroup sync.WaitGroup
+
+	waitgroup.Add(2)
+	go detector.validateApps(&waitgroup)
+	go detector.validateSpaces(&waitgroup)
+
+	waitgroup.Wait()
+}
+
+// ValidateApps performs CF App resource validation
+func (detector *Detector) validateApps(wg *sync.WaitGroup) {
+	defer wg.Done()
+
+	if !detector.config.AppConfig.Enabled {
+		return
+	}
+
+	deployedApps, err := detector.client.ListV3AppsByQuery(url.Values{})
+	if err != nil {
+		log.Printf("ERROR in ListApps() request: %s. Skipping check.", err)
+		failedAppChecks.Inc()
+		return
+	}
+	deployedAppEntries := toAppEntries(deployedApps)
+	unknownApps := appDifference(deployedAppEntries, detector.config.AppConfig.Apps)
+	missingApps := appDifference(detector.config.AppConfig.Apps, deployedAppEntries)
+
+	log.Printf("Unknown Apps Detected: %s", unknownApps)
+	log.Printf("Missing Apps Detected: %s", missingApps)
+	totalUnknownApps.Set(float64(len(unknownApps)))
+	totalMissingApps.Set(float64(len(missingApps)))
+	successfulAppChecks.Inc()
 }
 
-func (detector *Detector) validateApps() {
-	deployedApps := detector.viewer.V3Apps()
-	unknownApps := 0
-OUTER:
-	for _, deployedApp := range deployedApps {
-		for _, allowedApp := range detector.config.Apps {
-			if allowedApp.Name == deployedApp.Name {
-				continue OUTER
+func (detector *Detector) validateSpaces(wg *sync.WaitGroup) {
+	defer wg.Done()
+
+	if !detector.config.SpaceConfig.Enabled {
+		return
+	}
+
+	visibleSpaces, err := detector.client.ListSpacesByQuery(url.Values{})
+	if err != nil {
+		log.Printf("ERROR in ListSpaces request: %s. Skipping check.", err)
+		failedSpaceChecks.Inc()
+		return
+	}
+
+	for _, space := range visibleSpaces {
+		for _, spaceEntry := range detector.config.SpaceConfig.Spaces {
+			if space.Name == spaceEntry.Name && space.AllowSSH != spaceEntry.AllowSSH {
+				log.Printf("Misconfigured SSH access detected for space: %s. SSH access enabled: %v", space.Name, space.AllowSSH)
+				totalSpaceSSHViolations.Inc()
 			}
 		}
-		log.Printf("Unknown App Detected: %s", deployedApp.Name)
-		unknownApps++
 	}
-	totalUnknownApps.Set(float64(unknownApps))
+	successfulSpaceChecks.Inc()
+}
+
+// App difference returns a slice of unknown app names.
+// Given two AppEntry slices, returns the set difference of the two
+// slices (a - b). This can be logically used as follows:
+// unknownApps = (the set of deployed apps) - (the set of valid apps) OR
+// missingApps = (the set of valid apps) - (the set of deployed apps)
+func appDifference(deployed, valid []AppEntry) (diff []string) {
```

Comment:
Seems like the args are miss-named on this function since it used twice in opposite inputs

Reply:
Agreed. How does the following seem?
```suggestion
func appDifference(a, b []AppEntry) (diff []string) {
```